docs/madr/decisions/056-reachable-svc-mesh-service.md

```diff
@@ -0,0 +1,239 @@
+
+Universal needs another entry in a `Dataplane` object:
+
+```proto
+message TransparentProxying {
+...
+    message ReachableBackendRef {
+        // Type of the backend: MeshService or MeshExternalService
+        //  +required
+        string kind = 1;
+
+        // Name of the backend.
+        //  +optional
+        string name = 2;
+
+        // Namespace of the backend. Might be empty
+        //  +optional
+        string namespace = 3;
+
+        // Port of the backend.
+        //  +optional
+        google.protobuf.UInt32Value port = 4;
+
+        // Labels used to select backends
+        //  +optional
+        map<string, string> labels = 5;
+    }
+
+    // List of reachable backend refs via transparent proxy when running with 
+    // MeshExternalService and MeshService. Setting an explicit list
+    // can dramatically improve the performance of the mesh. If not specified,
+    // all services in the mesh are reachable.
+    repeated ReachableBackendRef reachable_backends_ref = 7;
+    }
+```
+
+Most of the fields are optional and may not be used simultaneously. We need to support the following cases:
+* When `name`, `namespace`, and `port` are defined(combination of them, all are not requried), you cannot use `labels`.
+* When `labels` are defined, you cannot use `name`, `namespace`, and `port`.
+
+Another thing that is required to do is add another fields to the `BackendRef` in Dataplane object: namespace and labels
+```proto
+      message BackendRef {
+        // Kind is a type of the object to target. Allowed: MeshService, MeshExternalService
+        string kind = 1;
+        // Name of the targeted object
+        string name = 2;
+        // Port of the targeted object. Required when kind is MeshService, MeshExternalService
+        uint32 port = 3;
+        // Namespace of the targeted object.
+        //  +optional
+        string namespace = 4;
+        // Labels of the targeted object
+        map<string, string> labels = 5;
+      }
+```
+
+This enhancement should make it easier to match the reachable services provided by the user with the generated or provided outbounds.
+
+### Autoreachable services
+
+From a code perspective, we used to build a map of supported tags: `kuma.io/service`, `k8s.kuma.io/namespace`, `k8s.kuma.io/service-name`, and `k8s.kuma.io/service-port`, and later match `MeshTrafficPermission` to dataplanes with these tags. Since the introduction of `MeshService` and `MeshExternalService`, we might not need to iterate over all dataplanes. Instead, we can match all `MeshService` and `MeshExternalService` entries. Based on this information, we know which dataplanes need to be configured. The only limitation is that `MeshService` doesn't carry all tags that a dataplane might have, which means we can support a limited scope of tags, similar to how we do with `kuma.io/service`.
```

Comment:
It's not clear from the description that to avoid iterating over ddps we can allow MeshSubet but *only* with `dpTags` selected by MeshService.

This seems ok, as it's very likely that if Service Owner has a "main application tag" like `app`, they will use the same thing in `MeshTrafficPermission`

Reply:
Tried to clarify this

Reply:
I think a simple algo definition would be helpful here to clarify how we want to do it. It's a bit handwavy right now.

It's something like
1) Take each MeshService
2) For each MeshService, take tags from dpTags (+kuma.io/zone, +kuma.io/origin) and build a fake "DPP" with one inbound
3) Execute MatchedPolicies on this "DDP" with a list of MeshTrafficPermission to get rules
4) Store rules for each MeshService.
5) When we build XDS config for client DPP and check if we should add a cluster/listener for given MeshService, execute rules for MeshService using client DPP.

Writing this clarifies why we can only support dpTags defined in MeshService.

Also for each MeshService we need to transform list of MeshTrafficPermission using `trimNotSupportedTags`. This time we need to pass the tags of fake DPP instead of static list of SupportedTags. I would not get into details of this func, just reference it.